### PR TITLE
feat(builder, server): Add Builds to Frontend and Packaged assets

### PR DESCRIPTION
**Pull request by ntindle**
### Background

We need to be able to serve a frontend and backend on the installed versions of AutoGPT Server. This fixes errors and adds that capability to the frontend and backend.

### Changes 🏗️

Unfortunately, lots of changes were required to get this to work.

First, `next build` runs `next lint,` so all the errors from `next lint` were fixed.
Next, the build path for the frontend was updated to be `/frontend` rather than `/example_files.`
Then, the frontend was correctly served, if available in the `/frontend.` The routes were also hooked to serve the frontend pages with a higher priority than the API endpoints with a higher priority than the static files.
Finally, the packaging script `setup.py` and the ci pipelines were updated.

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [x] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 1]
```diff
@@ -135,20 +140,50 @@
...
+        async def catch_others(extras_file: str):
+            # check if the file is directly in the frontend path folder to prevent directory traversal
+            files = os.listdir(get_frontend_path())
+            # detect the file type and return the appropriate response
+            if extras_file in files:
+                mime = mimetypes.guess_type(extras_file)[0]
+                if mime is not None and mime.startswith("image"):
+                    return FileResponse(get_frontend_path() / extras_file)
+                else:
+                    with open(get_frontend_path() / extras_file) as file:
```
[the remark this answers is not in the document]
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/server.py:233]: Followed the best practices linked here

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 2]
```diff
@@ -178,20 +182,63 @@ def run_service(self):
...
+                app=SPAStaticFiles(directory=get_frontend_path() / "_next", html=True),
+                name="frontend",
+            )
+
         @app.websocket("/ws")
         async def websocket_endpoint(websocket: WebSocket):  # type: ignore
             await self.websocket_router(websocket)
 
+        @app.get("/{extras_file:str}", response_class=HTMLResponse)
+        async def catch_others(extras_file: str):
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/server/server.py:219]: Why this needs to be handled manually and can't use app.mount to serve static files too ?

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 3]
```diff
@@ -178,20 +182,63 @@ def run_service(self):
...
+            # Do not allow directory traversal
+            if ".." in extras_file:
+                raise HTTPException(status_code=400, detail="Invalid file path")
+            if extras_file.count(".") > 1:
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/server/server.py:223]: If the file has no extension, can it use relative path ?
Can we use if not `os.path.isabs` to avoid the accessing relative path?

**Review by majdyz — DISMISSED** [review]

File: rnd/autogpt_server/autogpt_server/server/server.py (hunk 2 above)
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/server.py:219]: I think it could be but it would require the app router be mounted at /api/ and the frontend mounted at /frontend. IMO this is a better experience which mounts the pages on the root route by default

File: .github/workflows/autogpt-server-ci.yml [hunk 4]
```diff
@@ -144,6 +146,17 @@ jobs:
...
+      - uses: actions/checkout@v4
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '21'
+
+      - name: Install dependencies for autogpt_builder
+        working-directory: rnd/autogpt_builder
+        run: |
+          npm install
```
Comment (Pwuts) [inline .github/workflows/autogpt-server-ci.yml]:
```
yarn
```

File: .github/workflows/autogpt-server-ci.yml [hunk 5]
```diff
@@ -183,11 +196,27 @@ jobs:
...
 
+      - name: Install Prisma Dependencies
+        run: |
+          mkdir prisma
+          poetry run prisma py fetch --force
+          ls ./prisma
+          pwd
+        env:
+          PRISMA_BINARY_CACHE_DIR: "./prisma"
+          PRISMA_HOME_DIR: "./prisma"
```
Comment (Pwuts) [inline .github/workflows/autogpt-server-ci.yml]: why is this necessary?

File: rnd/autogpt_builder/next.config.js [hunk 6]
```diff
@@ -1,8 +1,17 @@
 const dotenv = require('dotenv');
```
Comment (Pwuts) [inline rnd/autogpt_builder/next.config.js]: this needs to be merged into `next.config.mjs`

File: rnd/autogpt_builder/next.config.js [hunk 7]
```diff
@@ -1,8 +1,17 @@
...
 dotenv.config();
 
-module.exports = {
+const nextConfig = {
+  output: 'export',
```
Comment (Pwuts) [inline rnd/autogpt_builder/next.config.js]: nit: isn't `dist` generally the output folder for TS/React/etc projects? I'd suggest using that for minimum confusion

File: rnd/autogpt_builder/src/app/build/page.tsx [hunk 8]
```diff
@@ -1,8 +1,21 @@
...
+  const searchParams = useSearchParams();
+  const flowID = searchParams.get("flowID") ?? undefined;
+
+  return (
+    <FlowEditor
+      className="flow-container w-full min-h-[75vh] border border-gray-300 dark:border-gray-700 rounded-lg"
+      flowID={flowID}
+    />
+  );
+}
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/build/page.tsx]: unnecessary abstraction imo

File: rnd/autogpt_builder/src/app/build/page.tsx [hunk 9]
```diff
@@ -31,13 +44,11 @@ export default function Home() {
...
-                className="flow-container w-full min-h-[75vh] border border-gray-300 dark:border-gray-700 rounded-lg"
-                flowID={useSearchParams().get("flowID") ?? undefined}
-              />
-          </div>
+      <div className="w-full flex justify-center mt-10">
+        <Suspense fallback={<div>Loading...</div>}>
+          <FlowEditorWrapper />
+        </Suspense>
       </div>
+    </div>
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/build/page.tsx:52]: changed indentation?

File: rnd/autogpt_builder/src/app/build/page.tsx [hunk 10]
```diff
@@ -31,13 +44,11 @@ export default function Home() {
...
   );
-}
+}
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/build/page.tsx]: missing newline

File: rnd/.gitignore [hunk 11]
```diff
@@ -0,0 +1 @@
+frontend/
```
Comment (Pwuts) [inline rnd/.gitignore:1]: * what's this for? 
* missing newline

File: rnd/autogpt_builder/src/app/monitor/page.tsx [hunk 12]
```diff
@@ -266,6 +266,7 @@ const AgentFlowList = (
   </Card>
 );
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/monitor/page.tsx:269]: ?

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 13]
```diff
@@ -121,7 +131,7 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
   const isHandleConnected = (key: string) => {
     return data.connections && data.connections.some((conn: any) => {
       if (typeof conn === 'string') {
-        const [source, target] = conn.split(' -> ');
+        const [ target] = conn.split(' -> ');
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx]: inconsistent spacing: `[ target]`

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 14]
```diff
@@ -168,11 +178,13 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
...
     }
 
-    const renderClickableInput = (displayValue: string) => (
-      <div className="clickable-input" onClick={() => handleInputClick(fullKey)}>
-        {displayValue}
-      </div>
-    );
+  const renderClickableInput = (displayValue: unknown): JSX.Element => (
+    <div className="clickable-input" onClick={() => handleInputClick(fullKey)}>
+      {typeof displayValue === 'string' || typeof displayValue === 'number' || typeof displayValue === 'boolean'
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx]:
```suggestion
      {["string", "number", "boolean"].includes(typeof displayValue)
```

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 15]
```diff
@@ -168,11 +178,13 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
...
+        ? String(displayValue)
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx]: for the three selected types, I think `JSON.stringify` does the same as `String(v)`/`v.toString()`

so then this whole thing can just be `JSON.stringify(displayValue)`

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 16]
```diff
@@ -168,11 +178,13 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
...
+        : JSON.stringify(displayValue)}
+    </div>
+  );
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx]: bad indentation

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 17]
```diff
@@ -297,7 +309,7 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
         return schema.enum ? (
           <div key={fullKey} className="input-container">
             <select
-              value={value || ''}
+              value={getSelectValue(value)}
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx:313]: are there currently any cases where this is necessary?

File: rnd/autogpt_builder/src/components/CustomNode.tsx [hunk 18]
```diff
@@ -337,7 +349,7 @@ const CustomNode: FC<NodeProps<CustomNodeData>> = ({ data, id }) => {
           <div key={fullKey} className="input-container">
             <input
               type="number"
-              value={value || ''}
+              value={typeof value === 'number' ? value : typeof value === 'string' ? value : ''}
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/CustomNode.tsx]:
```suggestion
              value={["number", "string"].includes(typeof value) ? value : ''}
```

File: rnd/autogpt_builder/src/components/Flow.tsx [hunk 19]
```diff
@@ -1,3 +1,4 @@
+// @ts-nocheck
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/Flow.tsx:1]: bad

File: rnd/autogpt_builder/src/components/Flow.tsx [hunk 20]
```diff
@@ -253,6 +255,7 @@ const FlowEditor: React.FC<{ flowID?: string; className?: string }> = ({
     })));
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/components/Flow.tsx:258]: ?

File: rnd/autogpt_server/autogpt_server/app.py [hunk 21]
```diff
@@ -27,6 +27,7 @@ def run_processes(processes: list[AppProcess], **kwargs):
 
 
 def main(**kwargs):
+    breakpoint()
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/app.py]: ?

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 22]
```diff
@@ -178,20 +182,63 @@ def run_service(self):
...
+                    return FileResponse(get_frontend_path() / extras_file)
+                else:
+                    with open(get_frontend_path() / extras_file) as file:
+                        return str(file.read())
+            raise HTTPException(status_code=400, detail="Invalid file path")
+
+        @app.get("/", response_class=HTMLResponse)
+        async def catch_root():
+            with open(get_frontend_path() / "build.html") as file:
+                return str(file.read())
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/server.py:240]: This shouldn't be necessary once `next.config.js` is merged into `next.config.mjs`, which contains a route for `/` -> `/build`.

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 23]
```diff
@@ -178,20 +182,63 @@ def run_service(self):
...
+                path="/_next",
+                app=SPAStaticFiles(directory=get_frontend_path() / "_next", html=True),
+                name="frontend",
+            )
+
         @app.websocket("/ws")
         async def websocket_endpoint(websocket: WebSocket):  # type: ignore
             await self.websocket_router(websocket)
 
+        @app.get("/{extras_file:str}", response_class=HTMLResponse)
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/server.py:218]: `HTMLResponse` also for CSS and JS assets?

File: rnd/autogpt_server/autogpt_server/server/server.py [hunk 24]
```diff
@@ -178,20 +182,63 @@ def run_service(self):
...
+        async def catch_others(extras_file: str):
+            # Do not allow directory traversal
+            if ".." in extras_file:
+                raise HTTPException(status_code=400, detail="Invalid file path")
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/server.py:222]: would `GET //home/users/nick` work?

File: rnd/autogpt_server/autogpt_server/util/data.py [hunk 25]
```diff
@@ -14,12 +14,12 @@ def get_config_path() -> pathlib.Path:
...
     if getattr(sys, "frozen", False):
         # The application is frozen
-        datadir = pathlib.Path(os.path.dirname(sys.executable)) / "example_files"
+        datadir = pathlib.Path(os.path.dirname(sys.executable)) / "frontend"
     else:
         # The application is not frozen
         # Change this bit to match where you store your data files:
         filedir = os.path.dirname(__file__)
-        datadir = pathlib.Path(filedir).parent.parent.parent / "example_files"
+        datadir = pathlib.Path(filedir).parent.parent.parent / "frontend"
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/util/data.py:22]:
```suggestion
        autogpt_server_dir = Path(__file__).parent.parent.parent
        frontend_dir = autogpt_server_dir.parent / "autogpt_builder" / "dist"
```

File: rnd/autogpt_server/setup.py [hunk 26]
```diff
@@ -57,6 +59,26 @@ def txt_to_rtf(input_file: Union[str, Path], output_file: Union[str, Path]) -> N
...
+# call npm run build in ../autogpt_builder
+
+
+os.system("npm run build --prefix ../autogpt_builder")
+
+# copy the ../autogpt_builder/out to frontend
+
+
+shutil.rmtree("../frontend", ignore_errors=True)
+shutil.copytree("../autogpt_builder/out", "../frontend")
```
Comment (Pwuts) [inline rnd/autogpt_server/setup.py:69]: `/out`? afaics this is currently set to `/output` (and I proposed `/dist`)

File: rnd/autogpt_server/setup.py [hunk 27]
```diff
@@ -57,6 +59,26 @@ def txt_to_rtf(input_file: Union[str, Path], output_file: Union[str, Path]) -> N
...
 
+# call npm run build in ../autogpt_builder
+
+
+os.system("npm run build --prefix ../autogpt_builder")
+
+# copy the ../autogpt_builder/out to frontend
+
+
+shutil.rmtree("../frontend", ignore_errors=True)
```
Comment (Pwuts) [inline rnd/autogpt_server/setup.py]:
```suggestion
# call npm run build in ../autogpt_builder
os.system("npm run build --prefix ../autogpt_builder")

# copy the ../autogpt_builder/out to frontend
shutil.rmtree("../frontend", ignore_errors=True)
```

File: rnd/autogpt_server/setup.py [hunk 28]
```diff
@@ -2,6 +2,8 @@
 from pathlib import Path
 from pkgutil import iter_modules
 from typing import Union
+import shutil
+import os
```
Comment (Pwuts) [inline rnd/autogpt_server/setup.py:6]: nit: import sorting (but we need to set up formatting anyway)

File: rnd/frontend/.gitkeep [hunk 29]
```diff
@@ -0,0 +1 @@
+# Copy your frontend builds here manually if you'd like to test the serving capibililties of the fast api server
```
Comment (Pwuts) [inline rnd/frontend/.gitkeep]: * missing newline
* s/capibilities/capabilities

File: rnd/autogpt_builder/src/app/monitor/page.tsx [hunk 30]
```diff
@@ -538,7 +539,7 @@ const FlowRunsTimeline = (
         tickFormatter={s => s > 90 ? `${Math.round(s / 60)}m` : `${s}s`}
       />
       <Tooltip
-        content={({ payload, label }) => {
+        content={({ payload}) => {
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/monitor/page.tsx]: inconsistent spacing: `{ payload}`

**Comment by Pwuts** [comment]
Also please add a scope to the PR title :)

File: .github/workflows/autogpt-server-ci.yml (hunk 5 above)
Comment (ntindle) [reply .github/workflows/autogpt-server-ci.yml]: The binaries aren’t being packaged right now if it’s not there. The other stuff is debug that needs removed

File: rnd/.gitignore (hunk 11 above)
Comment (ntindle) [reply rnd/.gitignore:1]: Staging area for cx-freeze to pull the built files from

File: rnd/autogpt_server/autogpt_server/app.py (hunk 21 above)
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/app.py]: Whoops, good find

File: .github/workflows/autogpt-server-ci.yml (hunk 5 above)
Comment (Pwuts) [reply .github/workflows/autogpt-server-ci.yml]: but what are the binaries for? `poetry install` should install everything that's needed for Prisma to work, right?

File: rnd/autogpt_server/autogpt_server/server/server.py (hunk 22 above)
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/server.py:240]: This is for the api side to do the same when serving it

File: .github/workflows/autogpt-server-ci.yml (hunk 5 above)
Comment (ntindle) [reply .github/workflows/autogpt-server-ci.yml]: Prisma installs it stuff outside site packages for some reason

File: rnd/autogpt_builder/next.config.js (hunk 7 above)
Comment (ntindle) [reply rnd/autogpt_builder/next.config.js]:
```
The type of build output.

undefined: The default build output, .next directory, that works with production mode next start or a hosting provider like Vercel
'standalone': A standalone build output, .next/standalone directory, that only includes necessary files/dependencies. Useful for self-hosting in a Docker container.
'export': An exported build output, out directory, that only includes static HTML/CSS/JS. Useful for self-hosting without a Node.js server.
```

File: rnd/autogpt_builder/next.config.js (hunk 6 above)
Comment (ntindle) [reply rnd/autogpt_builder/next.config.js]: Done

File: rnd/autogpt_builder/src/components/CustomNode.tsx (hunk 17 above)
Comment (ntindle) [reply rnd/autogpt_builder/src/components/CustomNode.tsx:313]: @aarushik93

File: rnd/autogpt_server/setup.py (hunk 26 above)
Comment (ntindle) [reply rnd/autogpt_server/setup.py:69]: that param doesn't do what it seems like :)

File: rnd/autogpt_builder/src/app/monitor/page.tsx (hunk 12 above)
Comment (aarushik93) [reply rnd/autogpt_builder/src/app/monitor/page.tsx:269]: just put this in for now til i do a proper linting and formatting PR to get the build working

File: rnd/autogpt_builder/src/components/Flow.tsx (hunk 19 above)
Comment (aarushik93) [reply rnd/autogpt_builder/src/components/Flow.tsx:1]: just put this in for now til i do a proper linting and formatting PR to get the build working

File: rnd/autogpt_builder/src/components/Flow.tsx (hunk 20 above)
Comment (aarushik93) [reply rnd/autogpt_builder/src/components/Flow.tsx:258]: just put this in for now til i do a proper linting and formatting PR to get the build working

File: rnd/autogpt_builder/src/components/CustomNode.tsx (hunk 17 above)
Comment (aarushik93) [reply rnd/autogpt_builder/src/components/CustomNode.tsx:313]: there's nothing in our nodes yet with different data types - but this fixes some of the TS check issues I was having - if there's a better solution lmk

File: .github/workflows/autogpt-server-ci.yml (hunk 5 above)
Comment (Pwuts) [reply .github/workflows/autogpt-server-ci.yml]: ack

File: rnd/autogpt_builder/src/components/CustomNode.tsx (hunk 17 above)
Comment (Pwuts) [reply rnd/autogpt_builder/src/components/CustomNode.tsx:313]: What was TypeScript's complaint?

File: rnd/autogpt_server/autogpt_server/server/server.py (hunk 22 above)
Comment (Pwuts) [reply rnd/autogpt_server/autogpt_server/server/server.py:240]: That's understood. Still, probably not necessary as Next.js will generate a route for `/` that handles the redirect. Having that logic in two different places - let alone two different codebases - is less than ideal. I also doubt that `/build` will be the default home page for much longer.

File: rnd/autogpt_server/setup.py (hunk 26 above)
Comment (Pwuts) [reply rnd/autogpt_server/setup.py:69]: Would you mind explaining? I looked it up in the docs and am none the wiser.

File: rnd/autogpt_builder/src/components/CustomNode.tsx (hunk 17 above)
Comment (aarushik93) [reply rnd/autogpt_builder/src/components/CustomNode.tsx:313]: `Type error: Type '{ [key: string]: any; }' is not assignable to type 'string | number | readonly string[] | undefined'.`

**Comment by ntindle** [comment]
Not going to do any longer
